BGZF RecursionError

rtf-const opened the pull request:
This pull request addresses issue  #1701 

<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file and understand that AppVeyor and
TravisCI will be used to confirm the Biopython unit tests and ``flake8`` style
checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

peterjc commented on Tests/test_bgzf.py:
```diff
@@ -416,6 +416,29 @@ def test_write_tell(self):
 
         h.close()
 
+    def test_bug_1701(self):
```
Can you rename that to something like ``test_many_blocks_in_one_read_call``?

mdshw5 replied:
Thanks for writing this test. I've had this in the back of my mind for almost a year and always procrastinated. This is really helpful!

peterjc commented:
This looks great - the fix is along the lines I was expecting/hoping would work, but you've made a nice reproducible test case which it the bit I was blocked on.

I noticed you haven't added your self to the NEWS and CONTRIB file, and don't put your name on your GitHub profile. If you want to be acknowledged as just @rtf-const that's fine - or leave it as is.

@mdshw5 is it easy for you to test this fix locally with pyfaidx, or would you prefer I merge it first?

mdshw5 commented:
I'll test this locally, but it looks like this will provide a fix for BGZF in pyfaidx, so feel free to merge.

peterjc commented:
Thanks @mdshw5 - I'll look forward to your test results 👍

mdshw5 commented on Bio/bgzf.py:
```diff
@@ -633,59 +633,55 @@ def read(self, size=-1):
+                # (lazy loading, don't load the next block unless we have too)
+                data = self._buffer[self._within_block_offset:self._within_block_offset + size]
+                self._within_block_offset += size
+                assert data  # Must be at least 1 byte
```
In the future it might be nice to have an `Exception` here in case the assertion is false. Something like `ValueError("Expected data at byte %d." % self.tell())`. For now I think this looks fine.

peterjc replied:
Fair point, yes.

rtf-const commented:
> If you want to be acknowledged as just @rtf-const that's fine - or leave it as is.

No, I have no special issues with that. I added my name to CONTRIB and NEWS and renamed the test method

peterjc commented:
Thank you 👍
